src/cache.rs

```diff
@@ -0,0 +1,632 @@
+    /// let mut cache = Vec::new();
+    /// ProguardCache::write(&mapping, &mut cache).unwrap();
+    /// let cache = ProguardCache::parse(&cache).unwrap();
+    ///
+    /// let mapped = cache.remap_class("a.a.a.a.c");
+    /// assert_eq!(mapped, Some("android.arch.core.executor.ArchTaskExecutor"));
+    /// ```
+    pub fn remap_class(&self, class: &str) -> Option<&'data str> {
+        let class = self.get_class(class)?;
+        self.read_string(class.original_name_offset).ok()
+    }
+
+    /// Remaps an obfuscated Class Method.
+    ///
+    /// The `class` argument has to be the fully-qualified obfuscated name of the
+    /// class, with its complete module prefix.
+    ///
+    /// If the `method` can be resolved unambiguously, it will be returned
+    /// alongside the remapped `class`, otherwise `None` is being returned.
+    pub fn remap_method(&self, class: &str, method: &str) -> Option<(&'data str, &'data str)> {
+        let class = self.get_class(class)?;
+        let members = self.get_class_members(class)?;
+        let mut iter = members.iter().filter(|m| {
+            let Ok(obfuscated_name) = self.read_string(m.obfuscated_name_offset) else {
+                return false;
+            };
+
+            obfuscated_name == method
+        });
+        let first = iter.next()?;
+
+        // We conservatively check that all the mappings point to the same method,
+        // as we don’t have line numbers to disambiguate.
+        // We could potentially skip inlined functions here, but lets rather be conservative.
+        let all_matching =
+            iter.all(|member| member.original_name_offset == first.original_name_offset);
+
+        if !all_matching {
+            return None;
+        }
+
+        let original_class = self.read_string(class.original_name_offset).ok()?;
+        let original_method = self.read_string(first.original_name_offset).ok()?;
+
+        Some((original_class, original_method))
+    }
+
+    /// Remaps a single Stackframe.
+    ///
+    /// Returns zero or more [`StackFrame`]s, based on the information in
+    /// the proguard mapping. This can return more than one frame in the case
+    /// of inlined functions. In that case, frames are sorted top to bottom.
+    pub fn remap_frame<'r: 'data>(
+        &'r self,
+        frame: &StackFrame<'data>,
+    ) -> RemappedFrameIter<'r, 'data> {
+        let Some(class) = self.get_class(frame.class) else {
+            return RemappedFrameIter::empty();
+        };
+
+        let mut frame = frame.clone();
+        let Ok(original_class) = self.read_string(class.original_name_offset) else {
+            return RemappedFrameIter::empty();
+        };
+
+        frame.class = original_class;
+
+        // The following if and else cases are very similar. The only difference
+        // is that if the frame contains parameter information, we use it in
+        // our comparisons (in addition to the method name).
```

Comment:
You could potentially solve this with generics, taking a `<T: PartialEq, F: Fn(&raw::Member) -> T>`, though a bit of duplication also doesn’t hurt.

Reply:
The problem is I need two closures, one for the binary search and one for the linear search.

Reply:
But shouldn’t `PartialEq` be a subset of `PartialOrd`? Meaning that you are only interested in `!Ordering::Equal`.

Reply:
You're right, that works. Good thinking.